bootstrap: fix and clean up bootstrapping at older versions

RaduBerinde opened the pull request:
Note: this PR is only for the top commit, the rest is #112456.

----

This commit fixes the logic around bootstrapping clusters at earlier
versions. Currently there is some special casing around the minimum
supported version which makes that work, but other backwards
compatible versions don't work correctly.  Specifically, after bumping
the version to 24.1, a `local-mixed-23.2` config for logic tests does
not work.

The core problem is that we are initializing the stores with a version
and then populating the initial system values at another version.

I am reasonably confident this was the root cause for #104994, where
we tried adding a 23.1 config while the min supported version was
22.2.

This fix makes the logic clearer and simpler. We can only bootstrap
the cluster at specific versions (the current version, and compatible
released versions). When the bootstrap version is overridden, we find
the latest version that we can initialize the cluster and use that for
both the stores and the initial values. Then we can upgrade to the
desired version from there.

This simplifies the tests: we no longer need
`BootstrapVersionKeyOverride` (which was confusing, given that we also
have `BinaryVersionOverride`).

Epic: none
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/112385)
<!-- Reviewable:end -->

stevendanna reviewed (approved):
Overall I really like this change, one less frustrating bit of confusion when trying to write a test that manipulates versions.

RaduBerinde commented:
TFTR! Will merge after I merge #112456 and rebase.

As a confirmation of the fix, the new `local-mixed-23` logictest config in #112271 now works.

RaduBerinde committed:
bootstrap: fix and clean up bootstrapping at older versions

This commit fixes the logic around bootstrapping clusters at earlier
versions. Currently there is some special casing around the minimum
supported version which makes that work, but other backwards
compatible versions don't work correctly.  Specifically, after bumping
the version to 24.1, a `local-mixed-23.2` config for logic tests does
not work.

The core problem is that we are initializing the stores with a version
and then populating the initial system values at another version.

I am reasonably confident this was the root cause for #104994, where
we tried adding a 23.1 config while the min supported version was
22.2.

This fix makes the logic clearer and simpler. We can only bootstrap
the cluster at specific versions (the current version, and compatible
released versions). When the bootstrap version is overridden, we find
the latest version that we can initialize the cluster and use that for
both the stores and the initial values. Then we can upgrade to the
desired version from there.

This simplifies the tests: we no longer need
`BootstrapVersionKeyOverride` (which was confusing, given that we also
have `BinaryVersionOverride`).

Epic: none
Release note: None

RaduBerinde commented:
`TestReplicateQueueRebalanceMultiStore` flake

bors r+
